Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,853 @@
+        
+        let lastWeekIdx = if daysLeftInLastWeek >= minimumDaysInFirstWeek {
+            weekRange.upperBound - 2
+        } else {
+            weekRange.upperBound - 1
+        }
+        
+        let weeks = weeks.map { weekIdx in
+           if weekIdx > 0 {
+               weekIdx - 1 + firstWeekIdx
+           } else {
+               lastWeekIdx + (weekIdx + 1)
+           }
+        }
+        
+        dates = dates.flatMap { date in
+            let week = calendar.component(.weekOfYear, from: date)
+            return weeks.compactMap { weekIdx in
+                let offset = weekIdx - week
+                return calendar.date(byAdding: .weekOfYear, value: offset, to: date)
+            }
+        }
+    }
+    
+    internal func _expandOrLimitMonths(dates: inout [Date], anchor: Date, action: ComponentAction) {
+        lazy var monthRange = calendar.range(of: .month, in: .year, for: anchor)!
+        let months = months.map { month in
+            if month.index > 0 {
+                return month
+            } else {
+                let newIndex = monthRange.upperBound + month.index
+                // The upper bound is the last month plus one. Subtracting 1 we get the last month
+                return Calendar.RecurrenceRule.Month(newIndex, isLeap: month.isLeap)
+            }
+        }
+        
+        if action == .limit {
+            dates = dates.filter {
+                let idx = calendar.component(.month, from: $0)
+                let isLeap = calendar._dateComponents([.month], from: $0).isLeapMonth
+                return months.contains {
+                    $0.index == idx && $0.isLeap == isLeap
+                }
+            }
+        } else {
+            let componentSet: Calendar.ComponentSet = [ .month, .isLeapMonth, .day, .hour, .minute, .second, .nanosecond ]
+            
+            let anchorComponents = calendar._dateComponents(componentSet, from: anchor)
+            let daysInYear = calendar.dateInterval(of: .year, for: anchor)!
+            // This is always the first expansion, so we can overwrite `dates`
+            dates = months.compactMap { month in
+                var components = anchorComponents
+                components.month = month.index
+                components.isLeapMonth = month.isLeap
+                return calendar.nextDate(after: daysInYear.start, matching: components, matchingPolicy: matchingPolicy)
+            }
+        }
+    }
+    internal func _expandOrLimitDaysOfTheMonth(dates: inout [Date], anchor: Date, action: ComponentAction) {
+        if action == .limit {
+            dates = dates.filter { date in
+                let day = calendar.component(.day, from: date)
+                var dayRange: Range<Int>? = nil
+                for dayOfMonth in daysOfTheMonth {
+                    if dayOfMonth > 0 {
+                        if dayOfMonth == day { return true }
+                    } else {
+                        if dayRange == nil {
```

Comment:
We can't have `lazy var` inside a closure for some reason, so we store the day range in an optional and calculate it the first time it is assigned.